### PR TITLE
added procedure docs

### PR DESCRIPTION
**Pull request by wouterdb**
# Description

* Short description here *

closes *Add ticket reference here*

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [ ] Attached issue to pull request
- [ ] Changelog entry
- [ ] Type annotations are present
- [ ] Code is clear and sufficiently documented
- [ ] No (preventable) type errors (check using make mypy or make mypy-diff)
- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

File: docs/lsm/embedded_entities/embedded_entities.rst [hunk 1]
```diff
@@ -236,38 +236,103 @@ For these more involved update scenarios we recommend updating the lifecycle spe
...
+  update_rejected                    active               -                     up                  -                        
+  update_acknowledged                active               -                     up                  -                        
+  update_inprogress                  active               rollback              -                   -                        
+  rollback                           active               candidate             -                   -                        
+====================== ============ ==================== ===================== =================== ========================= 
+
+
+6. Finally, translate to the state variables as follows:
+
+   1. on all non-validating state, current attributes should be ``active``
```
Comment (Hugo-Inmanta) [inline docs/lsm/embedded_entities/embedded_entities.rst:329]: This is not something that can be set in the State, right ? Is this more something to check the integrity of this table before using it to define the states ?
I don't understand the added value of this line.

Comment (Hugo-Inmanta) [inline docs/lsm/embedded_entities/embedded_entities.rst:329]: What about this, if it is more of a checklist type step: ? 
```suggestion
   1. For all non-validating states, double check that ``current_attributes==active``
```

File: docs/lsm/embedded_entities/embedded_entities.rst [hunk 2]
```diff
@@ -236,38 +236,103 @@ For these more involved update scenarios we recommend updating the lifecycle spe
...
+   2. on all non-validating state, set ``previous_attr_set_on_export`` to the value of ``previous attributes``
+   3. on all validating states ``current_attributes==state.validate_self``
+   4. on all validating states, ``previous_attr_set_on_validate`` to the value of ``previous attributes``
+   5. the same on the associated ``_failed`` states
```
Comment (Hugo-Inmanta) [inline docs/lsm/embedded_entities/embedded_entities.rst:333]:
```suggestion
   2. For all non-validating states, set ``previous_attr_set_on_export`` to the value of ``previous attributes``
   3. For all validating states, double check that ``current_attributes==state.validate_self``
   4. For all validating states, set ``previous_attr_set_on_validate`` to the value of ``previous attributes``
   5. Perform the same operations on the associated ``_failed`` states
```

File: docs/lsm/embedded_entities/embedded_entities.rst [hunk 3]
```diff
@@ -236,38 +236,103 @@ For these more involved update scenarios we recommend updating the lifecycle spe
...
+  rollback                                                                                                         
+====================== ============ ==================== ===================== ========= ========================= 
+
+3. Fill in states before the update and where we are actually performing the update or rollback. The ``current`` attribute will always be ``active`` and ``previous`` depends on the direction we are moving in. For the ``up`` state, we are no updating, so there is no ``previous``. For updates ``previous`` is always ``rollback`` (the old active state has been promoted to the ``rollback`` set) for a rollback scenarios, the ``previous`` attributes are always ``candidate``.
+
+====================== ============ ==================== ===================== ========= ========================= 
+  state                 validating   current attributes   previous attributes   is like   operation since is like  
+====================== ============ ==================== ===================== ========= ========================= 
+  up                                 active               -
```
Comment (Hugo-Inmanta) [inline docs/lsm/embedded_entities/embedded_entities.rst:288]: Renders like a bullet list with a single dash
```suggestion
  up                                 active               --                                                       
```

File: docs/lsm/embedded_entities/embedded_entities.rst [hunk 4]
```diff
@@ -236,38 +236,103 @@ For these more involved update scenarios we recommend updating the lifecycle spe
...
+  rollback                           active               candidate                                                
+====================== ============ ==================== ===================== ========= ========================= 
+
+
+4. for each state that remains, indicate which other state it pretends to be like: the state prior to the update or the state after the update. Also add all operations performed between the state and the state it is like.
+
+====================== ============ ==================== ===================== =================== ========================= 
+  state                 validating   current attributes   previous attributes   is like             operation since is like  
+====================== ============ ==================== ===================== =================== ========================= 
+  up                                 active               -                     -                   -
```
Comment (Hugo-Inmanta) [inline docs/lsm/embedded_entities/embedded_entities.rst:303]:
```suggestion
  up                                 active               --                    --                  --                       
```

File: docs/lsm/embedded_entities/embedded_entities.rst [hunk 5]
```diff
@@ -236,38 +236,103 @@ For these more involved update scenarios we recommend updating the lifecycle spe
...
+====================== ============ ==================== ===================== ========= ========================= 
+
+
+4. for each state that remains, indicate which other state it pretends to be like: the state prior to the update or the state after the update. Also add all operations performed between the state and the state it is like.
+
+====================== ============ ==================== ===================== =================== ========================= 
+  state                 validating   current attributes   previous attributes   is like             operation since is like  
+====================== ============ ==================== ===================== =================== ========================= 
+  up                                 active               -                     -                   -                        
+  update_start                                                                  up                  -
```
Comment (Hugo-Inmanta) [inline docs/lsm/embedded_entities/embedded_entities.rst:304]:
```suggestion
  update_start                                                                  up                  --                       
```

File: docs/lsm/embedded_entities/embedded_entities.rst [hunk 6]
```diff
@@ -236,38 +236,103 @@ For these more involved update scenarios we recommend updating the lifecycle spe
...
+  state                 validating   current attributes   previous attributes   is like             operation since is like  
+====================== ============ ==================== ===================== =================== ========================= 
+  up                                 active               -                     -                   -                        
+  update_start                                                                  up                  -                        
+  update_start          yes                                                     update_inprogress   promote/backwards        
+  update_rejected                                                               up                  -                        
+  update_acknowledged                                                           up                  -                        
+  update_inprogress                  active               rollback              -                   -                        
+  rollback                           active               candidate             -                   -
```
Comment (Hugo-Inmanta) [inline docs/lsm/embedded_entities/embedded_entities.rst:309]:
```suggestion
  update_rejected                                                               up                  --                       
  update_acknowledged                                                           up                  --                       
  update_inprogress                  active               rollback              --                  --                       
  rollback                           active               candidate             --                  --                       
```

File: docs/lsm/embedded_entities/embedded_entities.rst [hunk 7]
```diff
@@ -236,38 +236,103 @@ For these more involved update scenarios we recommend updating the lifecycle spe
...
+  state                 validating   current attributes   previous attributes   is like             operation since is like  
+====================== ============ ==================== ===================== =================== ========================= 
+  up                                 active               -                     -                   -                        
+  update_start                       active               -                     up                  -                        
+  update_start          yes          candidate            active                update_inprogress   promote/backwards        
+  update_rejected                    active               -                     up                  -                        
+  update_acknowledged                active               -                     up                  -                        
+  update_inprogress                  active               rollback              -                   -                        
+  rollback                           active               candidate             -                   -
```
Comment (Hugo-Inmanta) [inline docs/lsm/embedded_entities/embedded_entities.rst:323]:
```suggestion
  up                                 active               -                     -                   -                        
  update_start                       active               -                     up                  -                        
  update_start          yes          candidate            active                update_inprogress   promote/backwards        
  update_rejected                    active               --                    up                  --                       
  update_acknowledged                active               --                    up                  --                       
  update_inprogress                  active               rollback              --                  --                       
  rollback                           active               candidate             --                  --                       
```

**Review by Hugo-Inmanta — CHANGES_REQUESTED** [review]

File: docs/lsm/embedded_entities/embedded_entities.rst [hunk 8]
```diff
@@ -236,38 +236,103 @@ For these more involved update scenarios we recommend updating the lifecycle spe
...
-
-    * Previous set
-        - The previous set is the set defined by the ``previous_attr_set_on_export`` parameter.
+The following logic is used to determine which is the current and which is the previous attribute set
+
+================================= ================================== =================== 
+instance is being validate         previous                            current            
+================================= ================================== =================== 
+instance is being validate        ``previous_attr_set_on_validate``   ``validate_self``  
+instance is not being validated   ``previous_attr_set_on_export``     active_attributes
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply docs/lsm/embedded_entities/embedded_entities.rst:259]: it was, but perhaps it is not very clear/consistent?

File: docs/lsm/embedded_entities/embedded_entities.rst [hunk 9]
```diff
@@ -236,38 +236,103 @@ For these more involved update scenarios we recommend updating the lifecycle spe
...
+  update_start          yes                                                                                        
+  update_rejected                                                                                                  
+  update_acknowledged                                                                                              
+  update_inprogress                                                                                                
+  rollback                                                                                                         
+====================== ============ ==================== ===================== ========= ========================= 
+
+3. Fill in states before the update and where we are actually performing the update or rollback. The ``current`` attribute will always be ``active`` and ``previous`` depends on the direction we are moving in. For the ``up`` state, we are no updating, so there is no ``previous``. For updates ``previous`` is always ``rollback`` (the old active state has been promoted to the ``rollback`` set) for a rollback scenarios, the ``previous`` attributes are always ``candidate``.
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply docs/lsm/embedded_entities/embedded_entities.rst:283]:
```suggestion
3. Fill in states before the update and states where we are actually deploying the update. The ``current attributes`` will always be ``active`` and ``previous attributes`` depends on the direction we are moving in. For the ``up`` state, we are not updating, so there is no ``previous attributes``. For updates, ``previous attributes`` is always ``rollback`` (the old active set has been promoted to the ``rollback`` set). For rollback scenarios, the ``previous attributes`` are always ``candidate``.
```

File: docs/lsm/embedded_entities/embedded_entities.rst (hunk 1 above)
Comment (wouterdb) [reply docs/lsm/embedded_entities/embedded_entities.rst:329]: It is to check you have something that makes sense and that is also  its value. 

```suggestion
   1. on all non-validating state, current attributes should be ``active``. If it is not, either you made a mistake or the state tracking plugin can't be used for this feature and you will have to build one yourself. 
```

File: docs/lsm/embedded_entities/embedded_entities.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (Hugo-Inmanta) [reply docs/lsm/embedded_entities/embedded_entities.rst:259]: I see, what about `"active" attribute set`? (Not a strong opinion, can be ignored)

File: docs/lsm/embedded_entities/embedded_entities.rst [hunk 10]
```diff
@@ -236,38 +236,103 @@ For these more involved update scenarios we recommend updating the lifecycle spe
...
+  rollback                           active               candidate             --                  --                       
+====================== ============ ==================== ===================== =================== ========================= 
+
+5. Copy over the state of the ``is like`` column and apply the operations.
+
+====================== ============ ==================== ===================== =================== ========================= 
+  state                 validating   current attributes   previous attributes   is like             operation since is like  
+====================== ============ ==================== ===================== =================== ========================= 
+  up                                 active               -                     -                   -                        
+  update_start                       active               -                     up                  -
```
Comment (Hugo-Inmanta) [inline docs/lsm/embedded_entities/embedded_entities.rst:318]:
```suggestion
  up                                 active               --                    --                  --                       
  update_start                       active               --                    up                  --                       
```

File: docs/lsm/embedded_entities/embedded_entities.rst [hunk 11]
```diff
@@ -236,38 +236,103 @@ For these more involved update scenarios we recommend updating the lifecycle spe
...
+====================== ============ ==================== ===================== =================== ========================= 
+
+
+6. Finally, translate to the state variables as follows:
+
+   1. For all non-validating states, double check that ``current_attributes==active``
+   2. For all non-validating states, set ``previous_attr_set_on_export`` to the value of ``previous attributes``
+   3. For all validating states, double check that ``current_attributes==state.validate_self``
+   4. For all validating states, set ``previous_attr_set_on_validate`` to the value of ``previous attributes``
+   5. Perform the same operations on the associated ``_failed`` states
```
Comment (Hugo-Inmanta) [inline docs/lsm/embedded_entities/embedded_entities.rst:333]: I like the explicit meaning of the checks
```suggestion
   5. Perform the same operations on the associated ``_failed`` states

If any of the checks above failed, either you made a mistake or the state tracking plugin can't be used for this feature and you will have to build one yourself. 
```

**Review by Hugo-Inmanta — CHANGES_REQUESTED** [review]

File: docs/lsm/embedded_entities/embedded_entities.rst [hunk 12]
```diff
@@ -236,38 +236,103 @@ For these more involved update scenarios we recommend updating the lifecycle spe
...
+
+
+4. for each state that remains, indicate which other state it pretends to be like: the state prior to the update or the state after the update. Also add all operations performed between the state and the state it is like.
+
+====================== ============ ==================== ===================== =================== ========================= 
+  state                 validating   current attributes   previous attributes   is like             operation since is like  
+====================== ============ ==================== ===================== =================== ========================= 
+  up                                 active               -                     -                   -                        
+  update_start                                                                  up                  -                        
+  update_start          yes                                                     update_inprogress   promote/backwards
```
[the remark this answers is not in the document]
Comment (Hugo-Inmanta) [reply docs/lsm/embedded_entities/embedded_entities.rst:305]: I still don't fully grasp it but I think I have to live with it for now

**Review by Hugo-Inmanta — APPROVED** [review]

**Pull request merged**